File: kuksa-client/kuksa_client/cli_backend/grpc.py
########################################################################
# Copyright (c) 2022 Robert Bosch GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
########################################################################


import asyncio
import json
import pathlib
import queue
from typing import Any
from typing import Callable
from typing import Dict
from typing import Iterable
import uuid

from kuksa_client import cli_backend
import kuksa_client.grpc
import kuksa_client.grpc.aio
from kuksa_client.grpc import EntryUpdate


def callback_wrapper(callback: Callable[[str], None]) -> Callable[[Iterable[EntryUpdate]], None]:
    def wrapper(updates: Iterable[EntryUpdate]) -> None:
        callback(json.dumps([update.to_dict() for update in updates]))
    return wrapper


class Backend(cli_backend.Backend):
    def __init__(self, config):
        super().__init__(config)
        self.cacertificate = pathlib.Path(self.cacertificate)
        self.keyfile = pathlib.Path(self.keyfile)
        self.certificate = pathlib.Path(self.certificate)
        if self.tokenfile is not None:
            self.tokenfile = pathlib.Path(self.tokenfile)
            self.token = self.tokenfile.expanduser(
            ).read_text(encoding='utf-8').rstrip('\n')
        else:
            self.token = ""
        self.grpcConnected = False

        self.sendMsgQueue = queue.Queue()
        self.run = False

        self.AttrDict = {
            "value": (kuksa_client.grpc.Field.VALUE, kuksa_client.grpc.View.CURRENT_VALUE),
            "targetValue": (kuksa_client.grpc.Field.ACTUATOR_TARGET, kuksa_client.grpc.View.TARGET_VALUE),
            "metadata": (kuksa_client.grpc.Field.METADATA, kuksa_client.grpc.View.METADATA),
        }

    # Function to check connection status
    def checkConnection(self):
        if self.grpcConnected:
            return True
        return False

    # Function to stop the communication
    def stop(self):
        self.disconnect()
        self.run = False
        print("gRPC channel disconnected.")

    # Function to implement fetching of metadata
    def getMetaData(self, path, timeout=5):
        return self.getValue(path, "metadata", timeout)

    def updateMetaData(self, path, jsonStr, timeout=5):
        return self.setValue(path, jsonStr, "metadata", timeout)

    # Function to implement get
    def getValue(self, path, attribute="value", timeout=5):
        return self.getValues((path,), attribute, timeout)

    def getValues(self, paths, attribute="value", timeout=5):
        if attribute in self.AttrDict:
            field, view = self.AttrDict[attribute]
            entries = [kuksa_client.grpc.EntryRequest(
                path=path, view=view, fields=(field,)) for path in paths]
            requestArgs = {'entries': entries}
            return self._sendReceiveMsg(("get", requestArgs), timeout)

        return json.dumps({"error": "Invalid Attribute"})

    # Function to implement set
    def setValue(self, path, value, attribute="value", timeout=5):
        return self.setValues({path: value}, attribute, timeout)

    def setValues(self, updates: Dict[str, Any], attribute="value", timeout=5):
        if attribute in self.AttrDict:
            field, _ = self.AttrDict[attribute]
            entry_updates = []
            for path, value in updates.items():
                if field is kuksa_client.grpc.Field.VALUE:
                    entry = kuksa_client.grpc.DataEntry(
                        path=path, value=kuksa_client.grpc.Datapoint(value=value))
                elif field is kuksa_client.grpc.Field.ACTUATOR_TARGET:
                    entry = kuksa_client.grpc.DataEntry(
                        path=path, actuator_target=kuksa_client.grpc.Datapoint(
                            value=value),
                    )
                elif field is kuksa_client.grpc.Field.METADATA:
                    try:
                        metadata_dict = json.loads(value)
                    except json.JSONDecodeError:
                        return json.dumps({"error": "Metadata value needs to be a valid JSON object"})
                    entry = kuksa_client.grpc.DataEntry(
                        path=path, metadata=kuksa_client.grpc.Metadata.from_dict(
                            metadata_dict),
                    )
                entry_updates.append(kuksa_client.grpc.EntryUpdate(
                    entry=entry, fields=(field,)))
            requestArgs = {'updates': entry_updates}
            return self._sendReceiveMsg(("set", requestArgs), timeout)
        return json.dumps({"error": "Invalid Attribute"})

    # Function for authorization
    def authorize(self, tokenfile=None, timeout=5):
        if tokenfile is None:
            tokenfile = self.tokenfile
        tokenfile = pathlib.Path(tokenfile)
        requestArgs = {
            'token': tokenfile.expanduser().read_text(encoding='utf-8').rstrip('\n')
        }

        return self._sendReceiveMsg(("authorize", requestArgs), timeout)

    # Subscribe value changes of to a given path.
    # The given callback function will be called then, if the given path is updated.
    def subscribe(self, path, callback, attribute="value", timeout=5):
        return self.subscribeMultiple((path,), callback, attribute, timeout)

    def subscribeMultiple(self, paths, callback, attribute="value", timeout=5):
        if attribute in self.AttrDict:
            field, view = self.AttrDict[attribute]
            entries = [kuksa_client.grpc.SubscribeEntry(
                path=path, view=view, fields=(field,)) for path in paths]
            requestArgs = {
                'entries': entries,
                'callback': callback_wrapper(callback),
            }
            return self._sendReceiveMsg(("subscribe", requestArgs), timeout)

        return json.dumps({"error": "Invalid Attribute"})

    # Unsubscribe value changes of to a given path.
    # The subscription id from the response of the corresponding subscription request will be required
    def unsubscribe(self, sub_id, timeout=5):
        try:
            sub_id = uuid.UUID(sub_id)
        except ValueError as exc:
            return json.dumps({"error": str(exc)})
        requestArgs = {'subscription_id': sub_id}
        return self._sendReceiveMsg(("unsubscribe", requestArgs), timeout)

    def connect(self, timeout=5):
        requestArgs = {}
        return self._sendReceiveMsg(("connect", requestArgs), timeout)

    def disconnect(self, timeout=5):
        requestArgs = {}
        return self._sendReceiveMsg(("disconnect", requestArgs), timeout)

    def _sendReceiveMsg(self, req, timeout):
        (call, requestArgs) = req
        recvQueue = queue.Queue(maxsize=1)
        self.sendMsgQueue.put((call, requestArgs, recvQueue))
        try:
            resp, error = recvQueue.get(timeout=timeout)
            if error:
                respJson = json.dumps(error, indent=4)
            elif resp:
                respJson = json.dumps(resp, indent=4)
            else:
                respJson = "OK"
        except queue.Empty:
            respJson = json.dumps({"error": "Timeout"})
        return respJson

    # Async function to handle the gRPC calls
    async def _grpcHandler(self, vss_client):
        self.grpcConnected = True
        self.run = True
        subscriber_manager = kuksa_client.grpc.aio.SubscriberManager(
            vss_client)
        while self.run:
            try:
                (call, requestArgs, responseQueue) = self.sendMsgQueue.get_nowait()
            except queue.Empty:
                await asyncio.sleep(0.01)
                continue
            try:
                if call == "get":
                    resp = await vss_client.get(**requestArgs)
                    if resp is not None:
                        resp = [entry.to_dict() for entry in resp]
                        resp = resp[0] if len(resp) == 1 else resp
                elif call == "set":
                    resp = await vss_client.set(**requestArgs)
                elif call == "authorize":
                    resp = await vss_client.authorize(str(requestArgs["token"]))
                elif call == "subscribe":
                    callback = requestArgs.pop('callback')
                    subscriber_response_stream = vss_client.subscribe(
                        **requestArgs)
                    resp = await subscriber_manager.add_subscriber(subscriber_response_stream, callback)
                    resp = {"subscriptionId": str(resp)}
                elif call == "unsubscribe":
                    resp = await subscriber_manager.remove_subscriber(**requestArgs)
                elif call == "connect":
                    resp = await vss_client.connect()
                elif call == "disconnect":
                    resp = await vss_client.disconnect()
                else:
                    raise Exception("Not Implemented.")

                responseQueue.put((resp, None))
            except kuksa_client.grpc.VSSClientError as exc:
                responseQueue.put((None, exc.to_dict()))
            except ValueError as exc:
                responseQueue.put(
                    (None, {"error": "ValueError in casting the value."}))

        self.grpcConnected = False

    # Main loop for handling gRPC communication
    async def mainLoop(self):
        if self.insecure:

            async with kuksa_client.grpc.aio.VSSClient(self.serverIP, self.serverPort, token=self.token) as vss_client:
                print("gRPC channel connected.")
                await self._grpcHandler(vss_client)
        else:
            async with kuksa_client.grpc.aio.VSSClient(
                self.serverIP,
                self.serverPort,
                root_certificates=self.cacertificate,
                private_key=self.keyfile,
                certificate_chain=self.certificate,
                token=self.token
            ) as vss_client:
                print("Secure gRPC channel connected.")
                await self._grpcHandler(vss_client)

File: kuksa_go_client/kuksa_client/grpc.go
// Copyright 2022 Robert Bosch GmbH
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package kuksa_client

import (
	"context"
	"errors"
	"log"
	"os"
	"reflect"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
	"google.golang.org/grpc/metadata"

	grpcpb "github.com/eclipse/kuksa.val/kuksa_go_client/proto/kuksa/val/v1"
	v1 "github.com/eclipse/kuksa.val/kuksa_go_client/proto/kuksa/val/v1"
)

type FieldView struct {
	Field grpcpb.Field
	View  grpcpb.View
}

type ParseError struct{}

func (e ParseError) Error() string {
	return "parse error"
}

var dict map[string]FieldView

func (cg *KuksaClientCommGrpc) ConnectToKuksaVal() error {
	dict = map[string]FieldView{
		"value":       {grpcpb.Field_FIELD_VALUE, grpcpb.View_VIEW_CURRENT_VALUE},
		"targetValue": {grpcpb.Field_FIELD_ACTUATOR_TARGET, grpcpb.View_VIEW_TARGET_VALUE},
		"metadata":    {grpcpb.Field_FIELD_METADATA, grpcpb.View_VIEW_METADATA},
	}
	err := cg.startCommunication()
	return err
}

// Function to get attribute value from databroker
func (cg *KuksaClientCommGrpc) GetValueFromKuksaVal(path string, attr string) ([]interface{}, error) {

	// Contact the server and return the Value
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	ctx = metadata.AppendToOutgoingContext(ctx, "Authorization", cg.authorizationHeader)
	defer cancel()
	view := dict[attr].View
	fields := []grpcpb.Field{dict[attr].Field}
	entries := grpcpb.EntryRequest{Path: path, View: view, Fields: fields}
	request := []*grpcpb.EntryRequest{&entries}
	resp, err := cg.client.Get(ctx, &grpcpb.GetRequest{Entries: request})
	if err != nil {
		log.Fatalf("Get call failed: %v", err)
	}
	respEntries := resp.GetEntries()
	response := make([]interface{}, len(respEntries))
	for i, entry := range respEntries {
		response[i] = interface{}(entry)
	}
	return response, nil
}

func GetArrayFromInput[T any](input string) ([]T, error) {
	// Strip the brackets from the input
	input = strings.TrimSuffix(strings.TrimPrefix(input, "["), "]")
	
	// Split the input string into separate values
	// First alternative, not quotes including escaped double quote, ends at comma, single/double quote or EOL
	// Second group is double quoted string, may contain double quoted strings inside, ends at non-escaped
	// double quote
	pattern := `(?:\\"|\\'|[^"',])+|"(?:\\"|[^"])*"|'(?:\\'|[^'])*'`

	r := regexp.MustCompile(pattern)
	values := r.FindAllString(input, -1)

	// Parse each value as type T and append to the array
	array := make([]T, 0)
	for _, v := range values {

		v = strings.TrimSpace(v)
		var consider = len(v) > 0
		// make the ' or " disappear
		if len(v) > 1 && (v[0] == '"') && (v[len(v)-1] == '"')  {
			v = v[1:len(v)-1]
		}
		if len(v) > 1 && (v[0] == '\'') && (v[len(v)-1] == '\'')  {
			v = v[1:len(v)-1]
		}
		v = strings.ReplaceAll(v,"\\\"", "\"")
		v = strings.ReplaceAll(v,"\\'", "'")
		if consider {
			value, err := parseValue[T](v)
			if err != nil {
				return nil, ParseError{}
			}
			array = append(array,value)
		}
	}

	return array, nil
}

func parseValue[T any](value string) (T, error) {
	t := reflect.TypeOf((*T)(nil)).Elem()
	switch t.Kind() {
	case reflect.Bool:
		parsed, err := strconv.ParseBool(strings.TrimSpace(value))
		if err != nil {
			return reflect.Zero(t).Interface().(T), err
		}
		v := reflect.New(t).Elem().Interface().(T)
		reflect.ValueOf(&v).Elem().SetBool(parsed)
		return v, nil
	case reflect.Int32:
		parsed, err := strconv.ParseInt(strings.TrimSpace(value), 10, 32)
		if err != nil {
			return reflect.Zero(t).Interface().(T), err
		}
		v := reflect.New(t).Elem().Interface().(T)
		reflect.ValueOf(&v).Elem().SetInt(parsed)
		return v, nil
	case reflect.Int64:
		parsed, err := strconv.ParseInt(strings.TrimSpace(value), 10, 64)
		if err != nil {
			return reflect.Zero(t).Interface().(T), err
		}
		v := reflect.New(t).Elem().Interface().(T)
		reflect.ValueOf(&v).Elem().SetInt(parsed)
		return v, nil
	case reflect.Float32:
		parsed, err := strconv.ParseFloat(strings.TrimSpace(value), 32)
		if err != nil {
			return reflect.Zero(t).Interface().(T), err
		}
		v := reflect.New(t).Elem().Interface().(T)
		reflect.ValueOf(&v).Elem().SetFloat(parsed)
		return v, nil
	case reflect.Float64:
		parsed, err := strconv.ParseFloat(strings.TrimSpace(value), 64)
		if err != nil {
			return reflect.Zero(t).Interface().(T), err
		}
		v := reflect.New(t).Elem().Interface().(T)
		reflect.ValueOf(&v).Elem().SetFloat(parsed)
		return v, nil
	case reflect.Uint32:
		parsed, err := strconv.ParseUint(strings.TrimSpace(value), 10, 32)
		if err != nil {
			return reflect.Zero(t).Interface().(T), err
		}
		v := reflect.New(t).Elem().Interface().(T)
		reflect.ValueOf(&v).Elem().SetUint(parsed)
		return v, nil
	case reflect.Uint64:
		parsed, err := strconv.ParseUint(strings.TrimSpace(value), 10, 64)
		if err != nil {
			return reflect.Zero(t).Interface().(T), err
		}
		v := reflect.New(t).Elem().Interface().(T)
		reflect.ValueOf(&v).Elem().SetUint(parsed)
		return v, nil
	case reflect.String:
		v := reflect.New(t).Elem().Interface().(T)
		reflect.ValueOf(&v).Elem().SetString(value)
		return v, nil
	default:
		return reflect.Zero(t).Interface().(T), errors.New("unsupported type")
	}
}

func (cg *KuksaClientCommGrpc) SetValueFromKuksaVal(path string, value string, attr string) error {
	// Contact the server and set the Value
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	ctx = metadata.AppendToOutgoingContext(ctx, "Authorization", cg.authorizationHeader)
	defer cancel()
	entries, getErr := cg.GetValueFromKuksaVal(path, "metadata")
	if getErr != nil {
		return getErr
	}

	var datatype int32
	for _, entry := range entries {
		metadata := entry.(*v1.DataEntry).GetMetadata()
		datatype = int32(metadata.GetDataType().Number())
	}

	// because of the lack of int8/16 and uint8/16 there are multiple times Datapoint_U/Int32 in the switch case
	// the switch case needs to be modified every time you add a new datapoint or datatype
	// the solution has been chosen because the compiler needs to know the type of a variable at runtime
	// second option would have been to have a setValue function for every type
	// this option has been not chosen because firstly the user needs to know what datatype his path requires (user exxperience low)
	// and secondly this would have been not less code only would have saved a little bit of runtime
	// there are maybe other options
	var datapoint grpcpb.Datapoint
	switch datatype {
	case 1:
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_String_{String_: value}}
	case 2:
		boolean, convErr := strconv.ParseBool(value)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Bool{Bool: boolean}}
	case 3:
		int32_, convErr := strconv.ParseInt(value, 10, 8)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Int32{Int32: int32(int32_)}}
	case 4:
		int32_, convErr := strconv.ParseInt(value, 10, 16)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Int32{Int32: int32(int32_)}}
	case 5:
		int32_, convErr := strconv.ParseInt(value, 10, 32)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Int32{Int32: int32(int32_)}}
	case 6:
		int64_, convErr := strconv.ParseInt(value, 10, 64)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Int64{Int64: int64_}}
	case 7:
		uint32_, convErr := strconv.ParseUint(value, 10, 8)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Uint32{Uint32: uint32(uint32_)}}
	case 8:
		uint32_, convErr := strconv.ParseUint(value, 10, 16)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Uint32{Uint32: uint32(uint32_)}}
	case 9:
		uint32_, convErr := strconv.ParseUint(value, 10, 32)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Uint32{Uint32: uint32(uint32_)}}
	case 10:
		uint64_, convErr := strconv.ParseUint(value, 10, 64)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Uint64{Uint64: uint64_}}
	case 11:
		float_, convErr := strconv.ParseFloat(value, 32)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Float{Float: float32(float_)}}
	case 12:
		float_, convErr := strconv.ParseFloat(value, 64)
		if convErr != nil {
			return convErr
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Double{Double: float_}}
	case 20:
		array, err := GetArrayFromInput[string](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_StringArray{StringArray: &grpcpb.StringArray{Values: array}}}
	case 21:
		array, err := GetArrayFromInput[bool](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_BoolArray{BoolArray: &grpcpb.BoolArray{Values: array}}}
	case 22:
		array, err := GetArrayFromInput[int32](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Int32Array{Int32Array: &grpcpb.Int32Array{Values: array}}}
	case 23:
		array, err := GetArrayFromInput[int32](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Int32Array{Int32Array: &grpcpb.Int32Array{Values: array}}}
	case 24:
		array, err := GetArrayFromInput[int32](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Int32Array{Int32Array: &grpcpb.Int32Array{Values: array}}}
	case 25:
		array, err := GetArrayFromInput[int64](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Int64Array{Int64Array: &grpcpb.Int64Array{Values: array}}}
	case 26:
		array, err := GetArrayFromInput[uint32](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Uint32Array{Uint32Array: &grpcpb.Uint32Array{Values: array}}}
	case 27:
		array, err := GetArrayFromInput[uint32](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Uint32Array{Uint32Array: &grpcpb.Uint32Array{Values: array}}}
	case 28:
		array, err := GetArrayFromInput[uint32](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Uint32Array{Uint32Array: &grpcpb.Uint32Array{Values: array}}}
	case 29:
		array, err := GetArrayFromInput[uint64](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_Uint64Array{Uint64Array: &grpcpb.Uint64Array{Values: array}}}
	case 30:
		array, err := GetArrayFromInput[float32](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_FloatArray{FloatArray: &grpcpb.FloatArray{Values: array}}}
	case 31:
		array, err := GetArrayFromInput[float64](value)
		if err != nil {
			return err
		}
		datapoint = grpcpb.Datapoint{Value: &grpcpb.Datapoint_DoubleArray{DoubleArray: &grpcpb.DoubleArray{Values: array}}}
	}

	fields := []grpcpb.Field{dict[attr].Field}
	var entry grpcpb.DataEntry
	if attr == "value" {
		entry = grpcpb.DataEntry{Path: path, Value: &datapoint}
	} else if attr == "targetValue" {
		entry = grpcpb.DataEntry{Path: path, ActuatorTarget: &datapoint}
	}

	requestargs := []*grpcpb.EntryUpdate{}
	requestargs = append(requestargs, &grpcpb.EntryUpdate{Entry: &entry, Fields: fields})
	_, err := cg.client.Set(ctx, &grpcpb.SetRequest{Updates: requestargs})
	return err
}

func (cg *KuksaClientCommGrpc) PrintSubscriptionMessages(id string) error {
	for {
		channel := *cg.subsChannel[id]
		resp, err := channel.Recv()
		if err != nil {
			return err
		}
		updates := resp.GetUpdates()
		log.Println(updates)

		for _, update := range updates {
			var val string
			entry := update.GetEntry()
			switch cg.subsAttr {
			case "value":
				val = entry.GetValue().String()
			case "targetValue":
				val = entry.GetActuatorTarget().String()
			case "metadata":
				val = entry.GetMetadata().String()
			default:
				continue
			}
			log.Printf("Vehicle.Speed Subscribed: %s", val)
		}
	}
}

func (cg *KuksaClientCommGrpc) SubscribeFromKuksaVal(path string, attr string) (string, error) {
	// Contact the server and return SubscriptionId
	cg.subsAttr = attr
	ctx, cancel := context.WithCancel(context.Background())
	ctx = metadata.AppendToOutgoingContext(ctx, "Authorization", cg.authorizationHeader)
	id := uuid.New().String()
	cg.cancel[id] = &cancel
	view := dict[attr].View
	fields := []grpcpb.Field{dict[attr].Field}
	subEntry := []*grpcpb.SubscribeEntry{{Path: path, View: view, Fields: fields}}
	subReq := grpcpb.SubscribeRequest{Entries: subEntry}
	client, err := cg.client.Subscribe(ctx, &subReq)
	if err != nil {
		return "", err
	}
	cg.subscriptionStateMutex.Lock()
	cg.subsChannel[id] = &client
	cg.subscriptionStateMutex.Unlock()
	return id, nil
}

func (cg *KuksaClientCommGrpc) UnsubscribeFromKuksaVal(id string) error {
	cancel := *cg.cancel[id]
	cancel()
	client := *cg.subsChannel[id]
	client.CloseSend()

	return nil
}

// Function to authorize to kuksa.val server
func (cg *KuksaClientCommGrpc) AuthorizeKuksaValConn(TokenOrTokenfile string) error {
	var tokenString string
	// Data in config has precedence over hardcoded token
	if cg.Config.TokenOrTokenfile != ""{
		tokenString = cg.Config.TokenOrTokenfile
	}else{
		if TokenOrTokenfile == "" {
			return errors.New("No token given!")
		}
		tokenString = TokenOrTokenfile
	}
	
	log.Printf("Using token: %s", tokenString)

	info, err := os.Stat(tokenString)
	if err != nil {
		// the TokenOrTokenfile is read like its a token
	} else if info.Mode().IsRegular() {
		tokenByteString, err := os.ReadFile(tokenString)
		tokenString = string(tokenByteString)
		if err != nil {
			log.Fatal("Error reading token: ", err)
			return err
		}
	}
	cg.authorizationHeader = "Bearer " + tokenString
	return nil
}

// Function to get metadata from kuksa.val server
func (cg *KuksaClientCommGrpc) GetMetadataFromKuksaVal(path string) ([]interface{}, error) {
	metadata, err := cg.GetValueFromKuksaVal(path, "metadata")
	if err != nil {
		metadata = []interface{}{}
		return metadata, err
	}
	return metadata, nil
}
